students/dave_arasim/session05/mailroom3.py

```diff
@@ -0,0 +1,251 @@
+
+    # Build the Donor Dictionary from a text file
+    donor_dict = defaultdict(list)
+    build_donor_dict(donor_dict)
+
+    # Build switch_dict with a dict comprehension
+    switch_options = [1, 2, 3, 4]
+    switch_functions = [send_thank_you, create_report, send_all_thanks, quit_option_true]
+    switch_dict = {option: function for option, function in zip(switch_options, switch_functions)}
+
+    while not(quit_option):
+        clear_screen()
+
+        print('*'*10,'  Main Menu  ','*'*10)
+        print()
+        print('1) Send a Thank You to a single donor')
+        print('2) Create a Report')
+        print('3) Send Thank You letters to all donors')
+        print('4) Quit')
+        print()
+        print('Choose an option: ', end='')
+
+        user_choice = input()
+        print()
+
+        try:
+            user_choice = int(user_choice)
+        except ValueError:
+            print('Sorry, but that is not a number')
+        else:
+            try:
+                switch_dict.get(user_choice)(donor_dict)
+            except TypeError:
+                print('That option: ', user_choice, ' is out of range', sep='')
+
+        print()
+
+        if quit_option:
+            print('Quitting process...')
+        else:
+            print('<cr> to continue... ', end='')
+            cont_input = input()  # cont_input for pause (variable not used)
+
+    build_donor_text(donor_dict)  # Export the Donor Dictionary back to a text file
+
+
+def clear_screen():
+    '''Clear the screen based on operating system in use'''
+    global this_os
+
+    if this_os == 'Windows':
+        os.system('clear')
+    else:
+        os.system('cls')
+
+
+def quit_option_true(donor_dict):
+    '''Set the quit_option to true, so function can be selected from switch_dict'''
+    global quit_option
+    quit_option = True
+
+
+def build_donor_dict(donor_dict):
+    '''Build Donor Dictionary by extracting values from a text file'''
+    try:
+        with open('donor_dict.txt', 'r') as dictfile:
+            while True:
+                donor_dict_line = dictfile.readline()
+                if not donor_dict_line: break
+
+                donor_dict_name, donor_dict_amts = donor_dict_line.strip().split(':')
+
+                # Build donor_dict with a list comprehension
+                donor_dict[donor_dict_name] = [float(amts) for amts in donor_dict_amts.split(',')]
+
+        dictfile.close()
+    except FileNotFoundError:
+        print()
+        print('Donor database in donor_dict.txt not found.  Starting with new, empty database.')
+        print()
+
+
+def build_donor_text(donor_dict):
+    '''Export the Donor Dictionary data back to a text file'''
+    with open('donor_dict.txt', 'w') as dictfile:
+        for this_name, this_amts in donor_dict.items():
+            amts_str = ','.join(str(x) for x in this_amts) # Comprehension
+            this_line = this_name + ':' + amts_str + '\n'
+
+            dictfile.write(this_line)
```

Comment:
Another good place for `try`-`except` in case the file writing fails due to errors like `PermissionError` or `IOError`.